### DISCUSSION TITLE
Ability to not drop duplicate identical frames when saving with CGIF

### DISCUSSION THREAD

**Discussion by cakedan** [libvips/libvips#4219] [Ideas]
Currently, by default, CGIF will drop identical frames from a gif. I just figured this out after using it with libvips for months now 😆. Is it possible to add the ability to disable this option?

The option is `CGIF_GEN_KEEP_IDENT_FRAMES` found here https://github.com/dloebl/cgif/blob/c234bc27d4012548faa13798c203219b59b77b09/inc/cgif.h#L17C9-L17C35 used here https://github.com/dloebl/cgif/blob/c234bc27d4012548faa13798c203219b59b77b09/src/cgif.c#L497

Not sure if it goes in the frame config or cgif config found here https://github.com/libvips/libvips/blob/6c397fb7eea08f2f34af6745d65d9f5bc15dbf4f/libvips/foreign/cgifsave.c#L620

Maybe have it be its own parameter in `gifsave` like "keep_duplicate_frames` which would be by default False?

Would love this option as my discord bot kinda depends on duplicate frames not being dropped for some commands :D

**Comment by jcupitt** [comment]
Hi @cakedan,

Rather than encoding two frames the same, couldn't you just extend the duration of one frame?

**Reply by jcupitt** [reply]
Ah, I see cgif does this automatically:

> // if frame matches previous frame, drop it completely and sum the frame delay

Could you explain why this doesn't work? Is the GIF viewer not respecting the `duration` attribute?

**Reply by cakedan** [reply]
The use case here is that my bot does media manipulation, some of which randomly recolor each frame in different ways.

The main roadblock I found was that I added a command `.fadein` which uses ffmpeg's `fade` filter. By default I have it fadein the media in the first 0.5 seconds. I was trying to take single frame images and making a 0.5 second gif out of them with libvips before sending it to ffmpeg so it would have the duration to be able to fade in properly. Technically I can use ffmpeg to increase the duration but I'd rather use libvips as it's easier.

Other use cases is someone using a command like `.seesaw` which duplicates the frames, then they use something like `.glitch` which would randomly color each frame. The result of `.seesaw` becomes a 1 frame gif and confuses the users.

Its a compression setting that I'd like to be able to disable without having to drop down to a lower cgif, been using libvips for like 4 years now and just now realized this undocumented thing was happening (looks like since about November 2022 lol).

**Reply by jcupitt** [reply]
Ahhh I see. I agree, this would be useful for cases like that. Let's add (yet another) enhancement issue.

**Reply by jcupitt** [reply]
https://github.com/libvips/libvips/issues/4220
